### DISCUSSION TITLE
reducer .ifLet seems to be causing `unable to type-check this expression in reasonable time`

### DISCUSSION THREAD

**Discussion by tisohjung** [pointfreeco/swift-composable-architecture#2725] [Q&A] [answered]
### Description

Hello, I'd love to get any suggestions, thank you 🙏

I get type-check error in `swift-composable-architecture` version > `1.4.0`
if I downgrade to `1.3.0` the error is gone.
if I comment out `.ifLet` chaining, the error is gone. It works up to 7 subFeatures

Here is the sample project where you can see `main` branch doesn't build, but `feature/downgrade-tca` branch works
I'd love to get any suggestions, thanks
 
https://github.com/tisohjung/ifletstore/blob/ad815a88a1969e7a7365e05b808ccfcbc295e294/IfLetStore/Feature1/Feature1.swift#L54
```swift

    .ifLet(\.subFeature1, action: /Action.subFeature1) { SubFeature1() }
    .ifLet(\.subFeature2, action: /Action.subFeature2) { SubFeature2() }
    .ifLet(\.subFeature3, action: /Action.subFeature3) { SubFeature3() }
    .ifLet(\.subFeature4, action: /Action.subFeature4) { SubFeature4() }
    .ifLet(\.subFeature5, action: /Action.subFeature5) { SubFeature5() }
    .ifLet(\.subFeature6, action: /Action.subFeature6) { SubFeature6() }
    .ifLet(\.subFeature7, action: /Action.subFeature7) { SubFeature7() }
    .ifLet(\.subFeature8, action: /Action.subFeature8) { SubFeature8() }
    .ifLet(\.subFeature9, action: /Action.subFeature9) { SubFeature9() }
```

### Checklist

- [ ] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

App should build on the latest version or `main` branch of swift-composable-architecture

### Actual behavior

App build under version `1.3.0`, and can't build above `1.4.0`

error shows:
`Feature1.swift:7:48 The compiler is unable to type-check this expression in reasonable time; try breaking up the expression into distinct sub-expressions`

<img width="553" alt="image" src="https://github.com/pointfreeco/swift-composable-architecture/assets/467106/2411481f-d28f-4ea5-a31a-84435ce674ef">

### Steps to reproduce

if I comment out `.ifLet` chaining, the error is gone. it works up to 7 subFeatures

### The Composable Architecture version information

1.6.0

### Destination operating system

iOS 17.2

### Xcode version information

Version 15.2 (15C500b)

### Swift Compiler version information

```shell
swift-driver version: 1.87.3 Apple Swift version 5.9.2 (swiftlang-5.9.2.2.56 clang-1500.1.0.2.5)
Target: arm64-apple-macosx14.0
```

**Comment by stephencelis** [comment, marked as answer]
Hi @tisohjung! Version 1.4 introduced "case key path" overloads for `ifLet`, which unfortunately expands the type checking space enough to cause regressions like this. While we will be able to remove these overloads in 2.0, we cannot make breaking changes till then.

In the meantime, you can extract these helpers to a dedicated place to improve compile times:

```swift
extension Reducer where State == Feature1.State, Action == Feature1.Action {
  func subFeatures() -> some ReducerOf<Self> {
    self
      .ifLet(\.subFeature1, action: /Action.subFeature1) { SubFeature1() }
      .ifLet(\.subFeature2, action: /Action.subFeature2) { SubFeature2() }
      .ifLet(\.subFeature3, action: /Action.subFeature3) { SubFeature3() }
      .ifLet(\.subFeature4, action: /Action.subFeature4) { SubFeature4() }
      .ifLet(\.subFeature5, action: /Action.subFeature5) { SubFeature5() }
      .ifLet(\.subFeature6, action: /Action.subFeature6) { SubFeature6() }
      .ifLet(\.subFeature7, action: /Action.subFeature7) { SubFeature7() }
      .ifLet(\.subFeature8, action: /Action.subFeature8) { SubFeature8() }
      .ifLet(\.subFeature9, action: /Action.subFeature9) { SubFeature9() }
  }
}
```

And you can break these methods down into `subFeatures1`, `subFeatures2`, etc., if need be.

Since this is more of a compiler issue than a bug in the library, I'm going to convert to a discussion.

**Reply by tisohjung** [reply]
awesome thanks 👍 

fix of the sample project is in `feature/fix` branch for those interested.
https://github.com/tisohjung/ifletstore/tree/feature/fix
```swift
extension Reducer where State == Feature1.State, Action == Feature1.Action {
  func subFeatures1to5() -> some ReducerOf<Self> {
    self
      .ifLet(\.subFeature1, action: /Action.subFeature1) { SubFeature1() }
      .ifLet(\.subFeature2, action: /Action.subFeature2) { SubFeature2() }
      .ifLet(\.subFeature3, action: /Action.subFeature3) { SubFeature3() }
      .ifLet(\.subFeature4, action: /Action.subFeature4) { SubFeature4() }
      .ifLet(\.subFeature5, action: /Action.subFeature5) { SubFeature5() }
  }

  func subFeatures6to9() -> some ReducerOf<Self> {
    self
      .ifLet(\.subFeature6, action: /Action.subFeature6) { SubFeature6() }
      .ifLet(\.subFeature7, action: /Action.subFeature7) { SubFeature7() }
      .ifLet(\.subFeature8, action: /Action.subFeature8) { SubFeature8() }
      .ifLet(\.subFeature9, action: /Action.subFeature9) { SubFeature9() }
  }
}
```
